### PR TITLE
TAM Enhancement for Monitoring Microbursts - header files update.

### PR DESCRIPTION
**Pull request by VitalyVovnoboy**
One more try

### CODE REVIEW

File: inc/sai.h [hunk 1]
```diff
@@ -64,6 +64,8 @@
 #include "saivirtualrouter.h"
 #include "saivlan.h"
 #include "saiwred.h"
+#include "saiuburst.h"
+
```
Comment (kcudnik) [inline inc/sai.h]: extra empty line

File: inc/saiuburst.h [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+ *    CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
+ *    LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS
+ *    FOR A PARTICULAR PURPOSE, MERCHANTABLITY OR NON-INFRINGEMENT.
+ *
+ *    See the Apache Version 2.0 License for specific language governing
+ *    permissions and limitations under the License.
+ *
+ *    Microsoft would like to thank the following companies for their review and
+ *    assistance with these files: Intel Corporation, Mellanox Technologies Ltd,
+ *    Dell Products, L.P., Facebook, Inc, Marvell Ltd.
```
Comment (kcudnik) [inline inc/saiuburst.h]: added Marvell - you will need to change style.pm to accept new header and update all headers to support that

File: inc/saiuburst.h [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+    SAI_TAM_MICROBURST_STAT_CUSTOM_RANGE_BASE = 0x10000000
+
+} sai_tam_microburst_stat_t;
+
+/**
+ * @brief TAM Microburst Attributes.
+ */
+typedef enum _sai_tam_microburst_attr_t {
+	/**
+	* @brief Start of Attributes
```
Comment (kcudnik) [inline inc/saiuburst.h]: wrong comment alignment

File: inc/saiuburst.h [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+	* @flags MANDATORY_ON_CREATE
+	* @objects SAI_OBJECT_TYPE_TAM_STAT
+	*/
+	SAI_TAM_MICROBURST_ATTR_STATISTIC,
+
+	/**
+	* @brief Watermark Levels
+	*
+    * Breach high/low watermark level for this microburst statistic
+    * in number of bytes.
```
Comment (kcudnik) [inline inc/saiuburst.h]: bad alignment

File: inc/saiuburst.h [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+     *
+     * In the absence of a transporter, the default transporter will
+     * be used.
+     *
+     * @type sai_object_id_t
+     * @flags CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_TAM_TRANSPORTER
+     * @allownull true
+     * @default SAI_NULL_OBJECT_ID
+     */
```
Comment (kcudnik) [inline inc/saiuburst.h]: bad comment alignment

File: inc/saiuburst.h [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+	SAI_TAM_MICROBURST_ATTR_TRANSPORTER,
+
+	/**
+    * @brief Buffers/Statistics for inclusion in the snapshot 
+    *
```
Comment (kcudnik) [inline inc/saiuburst.h]: bad alignement

File: inc/saiuburst.h [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+	*/
+	SAI_TAM_MICROBURST_ATTR_STAT_TYPES,
+
+	/**
+	* @brief End of Attributes
+	*/
+	SAI_TAM_MICROBURST_ATTR_END,
+
+    /** Custom range base value */
+    SAI_TAM_MICROBURST_ATTR_CUSTOM_RANGE_START = 0x10000000,
```
Comment (kcudnik) [inline inc/saiuburst.h]: bad alignement

File: inc/saiuburst.h [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+/**
+ * @brief Set histogram attribute value(s)
+ *
+ * @param[in] tam_histogram_id Histogram object id
+ * @param[in] attr Attribute to set
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_set_tam_histogram_attribute_fn)(
+        _In_ sai_object_id_t        tam_histogram_id,
```
Comment (kcudnik) [inline inc/saiuburst.h]: too many spaces

File: inc/saiuburst.h [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+ *
+ * @param[in] tam_histogram_id histogram object id
+ * @param[in] attr_count Number of attributes
+ * @param[inout] attr_list Array of attributes
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_get_tam_histogram_attribute_fn)(
+        _In_ sai_object_id_t    tam_histogram_id,
+        _In_ uint32_t           attr_count,
```
Comment (kcudnik) [inline inc/saiuburst.h]: too many spaces

File: inc/saiuburst.h [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_get_tam_histogram_stats_fn)(
+        _In_ sai_object_id_t tam_histogram_id,
+        _Inout_ uint32_t *number_of_bins,
+        _Inout_ uint32_t *bin_values);
+
+typedef struct _sai_uburst_api_t
+{
+    sai_create_tam_microburst_fn            create_tam_microburst;
```
Comment (kcudnik) [inline inc/saiuburst.h]: bad alignment

File: inc/saiuburst.h [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+	sai_remove_tam_microburst_fn            remove_tam_microburst;
+	sai_set_tam_microburst_attribute_fn     set_tam_microburst_attribute;
+	sai_get_tam_microburst_attribute_fn     get_tam_microburst_attribute;
+	sai_create_tam_histogram_fn             create_tam_histogram;
+	sai_remove_tam_histogram_fn             remove_tam_histogram;
+	sai_set_tam_histogram_attribute_fn      set_tam_histogram_attribute;
+	sai_get_tam_histogram_attribute_fn      get_tam_histogram_attribute;
+	sai_get_tam_histogram_stats_fn          get_tam_histogram_stats;
```
Comment (kcudnik) [inline inc/saiuburst.h]: add similar method to clear stats ?

**Comment by kcudnik** [comment]
you can check all changes locally by executing "cd meta; make"
and then submit working solution

**Comment by kcudnik** [comment]
convert tabs to 4 spaces

File: inc/saiuburst.h [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+ * @brief Create and return a microburst object
+ *
+ * @param[out] tam_microburst_id Microburst object
+ * @param[in] attr_count Number of attributes
+ * @param[in] attr_list Array of sai_tam_microburst_attr_t attributes
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_create_tam_microburst_fn)(
+        _Out_ sai_object_id_t *tam_microburst_id,
```
Comment (kcudnik) [inline inc/saiuburst.h]: function signature is missing switch_id parameter, take a look at other "create" functions

File: inc/saiuburst.h [hunk 13]
```diff
@@ -0,0 +1,401 @@
...
+
+/**
+ * @brief Delete a specified microburst object
+ *
+ * @param[in] tam_microburst_id Microburst object to be removed.
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_remove_tam_microburst_fn)(
+        _In_ sai_object_id_t *tam_microburst_id);
```
Comment (kcudnik) [inline inc/saiuburst.h]: this should not be a pointer

File: inc/saiuburst.h [hunk 14]
```diff
@@ -0,0 +1,401 @@
+/**
+ * Copyright (c) 20XX Microsoft Open Technologies, Inc.
```
Comment (kcudnik) [inline inc/saiuburst.h]: XX is actually replacement, you can put correct year here like 2017 etc

File: inc/saiuburst.h [hunk 15]
```diff
@@ -0,0 +1,401 @@
...
+ *
+ * @param[in] tam_microburst_id Microburst object id
+ * @param[in] attr_count Number of attributes
+ * @param[in] attr_list Array of attributes
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_set_tam_microburst_attribute_fn)(
+        _In_ sai_object_id_t tam_microburst_id,
+        _In_ uint32_t attr_count,
```
Comment (kcudnik) [inline inc/saiuburst.h]: set function should take only object_id and single attr, without count

File: inc/saiuburst.h [hunk 16]
```diff
@@ -0,0 +1,401 @@
...
+{
+    sai_create_tam_microburst_fn            create_tam_microburst;
+    sai_remove_tam_microburst_fn            remove_tam_microburst;
+    sai_set_tam_microburst_attribute_fn     set_tam_microburst_attribute;
+    sai_get_tam_microburst_attribute_fn     get_tam_microburst_attribute;
+    sai_create_tam_histogram_fn             create_tam_histogram;
+    sai_remove_tam_histogram_fn             remove_tam_histogram;
+    sai_set_tam_histogram_attribute_fn      set_tam_histogram_attribute;
+    sai_get_tam_histogram_attribute_fn      get_tam_histogram_attribute;
+    sai_get_tam_histogram_stats_fn          get_tam_histogram_stats;
```
Comment (kcudnik) [inline inc/saiuburst.h]: if stats can be cleared, please also add  "sai_clear_tam_histogram_stats_fn" function

Comment (VitalyVovnoboy) [inline inc/saiuburst.h]: No need for clear function here.

File: inc/saiuburst.h (hunk 15 above)
Comment (VitalyVovnoboy) [inline inc/saiuburst.h]: Done

File: inc/saiuburst.h (hunk 14 above)
Comment (VitalyVovnoboy) [inline inc/saiuburst.h]: Done

File: inc/saiuburst.h (hunk 13 above)
Comment (VitalyVovnoboy) [inline inc/saiuburst.h]: Fixed

File: inc/saiuburst.h (hunk 12 above)
Comment (VitalyVovnoboy) [inline inc/saiuburst.h]: done

File: inc/saiuburst.h [hunk 17]
```diff
@@ -0,0 +1,401 @@
...
+     * @brief Histogram Bins Lower Boundaries
+     *
+     * List of lower boundary of each bin for this HISTOGRAM in
+     * number referred object units. The upper boundary of a bin is
+     * the lower boundary of next bin. The upper boundary of the
+     * last bin is infinity.
+     *
+     * @type sai_u32_list_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     * @default empty
```
Comment (kcudnik) [inline inc/saiuburst.h]: default is not expected, since this is mandatory on create, remove default tag

Comment (VitalyVovnoboy) [inline inc/saiuburst.h]: Done

**Comment by kcudnik** [comment]
retest this please

File: inc/saiuburst.h [hunk 18]
```diff
@@ -0,0 +1,400 @@
...
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_TAM
+     */
+    SAI_TAM_MICROBURST_ATTR_TAM_ID = SAI_TAM_MICROBURST_ATTR_START,
+
+    /**
+     * @brief Statistic for this microburst
+     *
+     * @type sai_object_id_t
+     * @flags MANDATORY_ON_CREATE
```
Comment (kcudnik) [inline inc/saiuburst.h]: missing flags create_only or create_and_set

Comment (VitalyVovnoboy) [inline inc/saiuburst.h]: Done

**Comment by lguohan** [comment]
retest this please

File: inc/saiuburst.h [hunk 19]
```diff
@@ -0,0 +1,400 @@
...
+    SAI_TAM_HISTOGRAM_ATTR_TAM_ID = SAI_TAM_HISTOGRAM_ATTR_START,
+
+    /**
+     * @brief Buffers/Statistics for inclusion in histogram
+     *
+     * Specifies the Statistics/Types for a histogram.
+     *
+     * @type sai_object_list_t
+     * @flags CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_TAM_STAT, SAI_OBJECT_TYPE_TAM_MICROBURST
```
Comment (kcudnik) [inline inc/saiuburst.h]: is that correct that you can have multiple objects types on one list?

File: inc/saiuburst.h [hunk 20]
```diff
@@ -0,0 +1,400 @@
...
+     * @brief Buffers/Statistics for inclusion in histogram
+     *
+     * Specifies the Statistics/Types for a histogram.
+     *
+     * @type sai_object_list_t
+     * @flags CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_TAM_MICROBURST
+     * @default empty
+     */
+    SAI_TAM_HISTOGRAM_ATTR_STAT_TYPE,
```
Comment (kcudnik) [inline inc/saiuburst.h]: "type" would suggest that this attribute should be enum and not object_list, can you explain why is that ?

**Pull request merged**